### PR TITLE
Add Multihash for NMT node data

### LINKED ISSUE #142: nmt-rs Multihash compatibility
subtask of #133 

Custom multihash + nmt-rs compatibility https://github.com/celestiaorg/celestia-node/blob/34775bbf11e102f4a2457e09acdfc83c288ded73/share/ipld/nmt.go#L61-L68

### PR DESCRIPTION
**Pull request by fl0rek**
closes #142

### CODE REVIEW

File: types/src/multihash.rs [hunk 1]
```diff
@@ -0,0 +1,214 @@
+use multihash_derive::{Hasher, MultihashDigest};
+use nmt_rs::simple_merkle::tree::MerkleHash;
+use nmt_rs::NamespaceMerkleHasher;
+
+use crate::byzantine::MULTIHASH_SHA256_NAMESPACE_FLAGGED_CODE;
```
Comment (zvolin) [inline types/src/multihash.rs:5]: can we move this constant and other multihash related stuff there?

File: types/src/multihash.rs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+use crate::nmt::{
+    Namespace, NamespacedHashExt, NamespacedSha2Hasher, NAMESPACED_HASH_SIZE, NS_SIZE,
+};
+use crate::{Error, Result};
+
+pub type NamespacedHash = nmt_rs::NamespacedHash<NS_SIZE>;
```
Comment (zvolin) [inline types/src/multihash.rs:11]: can we move this alias to nmt mod and import it there?

File: types/src/multihash.rs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    hash: Option<NamespacedHash>,
+    hash_buf: [u8; NAMESPACED_HASH_SIZE],
+}
+
+impl Sha256NamespaceFlaggedHasher {
+    fn hash_data(&mut self, data: &[u8]) -> Result<()> {
+        let hash = if data.len() == NAMESPACED_HASH_SIZE * 2 {
+            let (left, right) = data.split_at(NAMESPACED_HASH_SIZE);
+            self.hasher
+                .hash_nodes(&left.try_into()?, &right.try_into()?)
```
Comment (zvolin) [inline types/src/multihash.rs:24]: after the len check we can just unwrap there

Comment (zvolin) [reply types/src/multihash.rs:24]: actually after we unwrap there, this method doesn't need to return the result

File: types/src/multihash.rs [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+}
+
+impl MultihashDigestExt for Code {
+    fn digest_leaf(&self, ns: &Namespace, data: &[u8]) -> Result<Multihash> {
+        let namespaced_data = [ns.as_bytes(), data].concat();
+
+        Ok(self.digest(&namespaced_data))
+    }
+
+    fn digest_inner(&self, left: &NamespacedHash, right: &NamespacedHash) -> Result<Multihash> {
```
Comment (zvolin) [inline types/src/multihash.rs:88]: can we do `digest_nodes`? I understand why digest_inner but it also can be understood as internal method that is used by others

File: types/src/multihash.rs [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        Self {
+            hasher: NamespacedSha2Hasher::with_ignore_max_ns(true),
+            hash: None,
+            hash_buf: [0u8; NAMESPACED_HASH_SIZE],
+        }
+    }
+}
+
+/// Outputs Multihashes, may panic when running `digest`. See `MultihashDigestExt` trait methods
+/// for safe alternative.
```
Comment (zvolin) [inline types/src/multihash.rs:68]: could you do `# Panics` section?

**Review by zvolin — COMMENTED** [review]
looks nice, left some ideas

File: types/src/multihash.rs (hunk 1 above)
Comment (fl0rek) [reply types/src/multihash.rs]: right, moved 👍

File: types/src/multihash.rs (hunk 2 above)
Comment (fl0rek) [reply types/src/multihash.rs]: Good idea, moved

File: types/src/multihash.rs (hunk 3 above)
Comment (fl0rek) [reply types/src/multihash.rs]: True - switched to `unwrap` and removed the `Result`.

File: types/src/multihash.rs (hunk 4 above)
Comment (fl0rek) [reply types/src/multihash.rs]: changed 👍

File: types/src/multihash.rs (hunk 5 above)
Comment (fl0rek) [reply types/src/multihash.rs]: Added note when provided `digest` can panic

**Comment by Wondertan** [comment]
Is this ipldv2 mh?

File: types/src/axis.rs [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+    }
+}
+
+impl HasCid<AXIS_ID_SIZE> for AxisId {
+    fn codec() -> u64 {
+        AXIS_ID_CODEC
+    }
+}
+
+impl<const S: usize> TryFrom<CidGeneric<S>> for AxisId {
```
Comment (oblique) [inline types/src/axis.rs:122]: Should this be `CidGeneric<AXIS_ID_SIZE>` instead?

File: types/src/sample.rs [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+
+        Ok(SampleId {
+            axis: AxisId::new(axis_type, axis_index, dah, block_height)?,
+            index: sample_index
+                .try_into()
+                .map_err(|_| Error::EdsIndexOutOfRange(sample_index))?,
+        })
+    }
+
+    fn to_bytes(&self, bytes: &mut BytesMut) {
```
Comment (oblique) [inline types/src/sample.rs:49]: `encode`

File: types/src/sample.rs [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+        // TODO: avoid alloc?
+        self.axis.to_bytes(bytes);
+        bytes.put_u16_le(self.index);
+    }
+
+    fn from_bytes(buffer: &[u8; SAMPLE_ID_SIZE]) -> Result<Self> {
```
Comment (oblique) [inline types/src/sample.rs:55]: `decode` and a slice

File: types/src/sample.rs [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+impl HasMultihash<SAMPLE_ID_SIZE> for SampleId {
+    fn multihash(&self) -> Result<Multihash<SAMPLE_ID_SIZE>> {
+        let mut bytes = BytesMut::with_capacity(Self::size());
+
+        self.to_bytes(&mut bytes);
+
+        Ok(Multihash::<SAMPLE_ID_SIZE>::wrap(
+            SAMPLE_ID_MULTIHASH_CODE,
+            &bytes[..],
+        )?)
```
Comment (oblique) [inline types/src/sample.rs:74]:
```suggestion
        Ok(Multihash::<SAMPLE_ID_SIZE>::wrap(
            SAMPLE_ID_MULTIHASH_CODE,
            &bytes[..],
        ).unwrap())
```

File: types/src/sample.rs [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+    }
+}
+
+impl HasCid<SAMPLE_ID_SIZE> for SampleId {
+    fn codec() -> u64 {
+        SAMPLE_ID_CODEC
+    }
+}
+
+impl<const S: usize> TryFrom<CidGeneric<S>> for SampleId {
```
Comment (oblique) [inline types/src/sample.rs:84]: Shouldn't this be `CidGeneric<SAMPLE_ID_SIZE>`?

File: types/src/multihash.rs [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+use nmt_rs::simple_merkle::tree::MerkleHash;
+use nmt_rs::NamespaceMerkleHasher;
+use tendermint::hash::SHA256_HASH_SIZE;
+
+use crate::nmt::{Namespace, NamespacedHash, NamespacedHashExt, NamespacedSha2Hasher, NS_SIZE};
+use crate::{Error, Result};
+
+pub const MULTIHASH_NMT_CODEC_CODE: u64 = 0x7700;
+pub const MULTIHASH_SHA256_NAMESPACE_FLAGGED_CODE: u64 = 0x7701;
+pub const MULTIHASH_SHA256_NAMESPACE_FLAGGED_SIZE: usize = 2 * NS_SIZE + SHA256_HASH_SIZE;
```
Comment (zvolin) [inline types/src/multihash.rs:12]: those feel inconsistent with `sample` and `axis` which don't have `MULTIHASH_` prefixes

**Review by zvolin — COMMENTED** [review]
Left some minor things but LGTM in general

File: types/src/axis.rs [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+    Row = 0,
+    Col,
+}
+
+impl TryFrom<i32> for AxisType {
+    type Error = Error;
+
+    fn try_from(value: i32) -> StdResult<Self, Self::Error> {
+        FromPrimitive::from_i32(value).ok_or(Error::InvalidAxis(value))
+    }
```
[the remark this answers is not in the document]
Comment (fl0rek) [reply types/src/axis.rs]: Agreed, removed the depedecies and done conversion by hand.

File: types/src/axis.rs [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+use crate::DataAvailabilityHeader;
+use crate::{Error, Result};
+
+const AXIS_ID_SIZE: usize = AxisId::size();
+pub const AXIS_ID_MULTIHASH_CODE: u64 = 0x7811;
+pub const AXIS_ID_CODEC: u64 = 0x7810;
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, FromPrimitive)]
+#[repr(u8)]
+pub enum AxisType {
```
[the remark this answers is not in the document]
Comment (fl0rek) [reply types/src/axis.rs:20]: I don't think we'll use naked `type` anywhere here, just `AxisType` - this matches with what golang is calling it, so i thought I'd be worth it to keep it unchanged. Wdyt?

File: types/src/axis.rs [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+        if block_height == 0 {
+            return Err(Error::ZeroBlockHeight);
+        }
+
+        let Some(dah_root) = (match axis_type {
+            AxisType::Row => dah.row_root(index),
+            AxisType::Col => dah.column_root(index),
+        }) else {
+            return Err(Error::EdsIndexOutOfRange(index));
+        };
```
[the remark this answers is not in the document]
Comment (fl0rek) [reply types/src/axis.rs]: Fixed, went with the latter. 👍

File: types/src/axis.rs [hunk 15]
```diff
@@ -0,0 +1,298 @@
...
+    }
+
+    pub const fn size() -> usize {
+        // size of:
+        // u8 + u16 + [u8; 32] + u64
+        //  1 +  2  +    32    +  8
+        43
+    }
+
+    pub(crate) fn to_bytes(&self, bytes: &mut BytesMut) {
```
[the remark this answers is not in the document]
Comment (fl0rek) [reply types/src/axis.rs]: It originally did, but working on existing buffer seemed more appropriate as more types were added. Changed to `encode`/`decode` here and later 👍

File: types/src/multihash.rs (hunk 11 above)
Comment (fl0rek) [reply types/src/multihash.rs:12]: true, renamed these to match convention better 👍

File: types/src/multihash.rs [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+
+pub trait HasCid<const S: usize>: HasMultihash<S> {
+    fn cid_v1(&self) -> Result<CidGeneric<S>> {
+        Ok(CidGeneric::<S>::new_v1(Self::codec(), self.multihash()?))
+    }
+
+    fn codec() -> u64;
+}
+
+impl HasMultihash<MULTIHASH_SHA256_NAMESPACE_FLAGGED_SIZE> for (&Namespace, &[u8]) {
```
[the remark this answers is not in the document]
Comment (fl0rek) [reply types/src/multihash.rs:26]: I like the idea, especially now that we have a range of hashes - moved 👍

File: types/src/axis.rs [hunk 17]
```diff
@@ -0,0 +1,298 @@
...
+        bytes.put_u8(self.axis_type as u8);
+        bytes.put_u16_le(self.index);
+        bytes.put(&self.hash[..]);
+        bytes.put_u64_le(self.block_height);
+    }
+
+    pub(crate) fn from_bytes(buffer: &[u8; AXIS_ID_SIZE]) -> Result<Self> {
```
[the remark this answers is not in the document]
Comment (fl0rek) [reply types/src/axis.rs]: Nice it simplified the caler side a bit 👌

File: types/src/axis.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (oblique) [reply types/src/axis.rs:21]: it is mostly about variable names and arguments: `axis_type: AxisType` vs `kind: AxisKind`

I'm ok with both. I prefer slightly the `kind`.

**Review by oblique — APPROVED** [review]
lgtm

**Review by zvolin — APPROVED** [review]

**Pull request merged**